File: code/backend/php/src/Core/Domain/Story/StoryValidationReport.php
<?php

declare(strict_types=1);

namespace Games\Paths\Core\Domain\Story;

/**
 * StoryValidationReport - accumulates {@see StoryValidationError}s (Step 22).
 */
final class StoryValidationReport
{
    /** @var StoryValidationError[] */
    private array $errors = [];

    public function add(string $rule, string $entityType, ?string $entityId, ?string $field, string $message): void
    {
        $this->errors[] = new StoryValidationError($rule, $entityType, $entityId, $field, $message);
    }

    public function isValid(): bool
    {
        return count($this->errors) === 0;
    }

    /** @return StoryValidationError[] */
    public function getErrors(): array
    {
        return $this->errors;
    }

    public function summary(): string
    {
        if (empty($this->errors)) {
            return 'story is valid';
        }
        $messages = array_map(fn (StoryValidationError $e) => $e->message, array_slice($this->errors, 0, 5));
        $head = implode('; ', $messages);
        return count($this->errors) <= 5 ? $head : $head . '; (+' . (count($this->errors) - 5) . ' more)';
    }

    public function toArray(): array
    {
        return [
            'valid' => $this->isValid(),
            'count' => count($this->errors),
            'errors' => array_map(fn (StoryValidationError $e) => $e->toArray(), $this->errors),
        ];
    }
}

File: code/backend/java/core/src/main/java/games/paths/core/model/story/StoryValidationReport.java
package games.paths.core.model.story;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

/**
 * StoryValidationReport - accumulates {@link StoryValidationError}s produced while
 * validating a story's referential integrity and domain rules (Step 22).
 *
 * <p>A report is {@link #isValid() valid} when it holds no errors.</p>
 */
public class StoryValidationReport {

    private final List<StoryValidationError> errors = new ArrayList<>();

    /** Adds an error to the report. */
    public void add(String rule, String entityType, String entityId, String field, String message) {
        errors.add(new StoryValidationError(rule, entityType, entityId, field, message));
    }

    /** @return true when no errors were recorded. */
    public boolean isValid() {
        return errors.isEmpty();
    }

    /** @return an immutable view of the recorded errors. */
    public List<StoryValidationError> getErrors() {
        return Collections.unmodifiableList(errors);
    }

    /** @return the number of recorded errors. */
    public int size() {
        return errors.size();
    }

    /** @return a compact one-line summary listing up to the first five messages. */
    public String summary() {
        if (errors.isEmpty()) {
            return "story is valid";
        }
        String head = errors.stream()
                .limit(5)
                .map(StoryValidationError::message)
                .collect(Collectors.joining("; "));
        return errors.size() <= 5
                ? head
                : head + "; (+" + (errors.size() - 5) + " more)";
    }
}
